mesonbuild/envconfig.py

```diff
@@ -86,8 +86,10 @@ class MesonConfigFile:
     @classmethod
     def from_config_parser(cls, parser: configparser.ConfigParser) -> T.Dict[str, T.Dict[str, T.Dict[str, str]]]:
         out = {}
+        constants = {'true': True, 'false': False}
```

Comment:
Don't we need to parse the constants section first to avoid a race between reading constants and evaluating sections? I think we need a test where `[constants]` isn't the first section.

Actually, taking that further, don't we need to read constants from each machine file, in order, before parsing any other sections in any other cross file to make sure that layering/overriding works correctly? We probably need a test for that too.

Reply:
On point 2 I think the configparser will actually handle that, but a test wouldn't be the worst thing to have :)

Reply:
In the doc I wrote:
```
Entries defined in the `[constants]` section can be used in any other section
after its definition (i.e. should usually be first section)
```
The idea is you could write:
```ini
[host_machine]
# constants can't be used here
system = 'linux'
...
[constants]
path='/foo'
...
[binaries]
# constants can be used here
c = path + '/gcc'
...
```
But I agree that wouldn't make much sense. I'm also not sure the ordering of sections is guaranteed in case for example the first file does not have [constants] section and 2nd file has [constants] is first section... I think the easiest solution here is to guarantee we parse [constants] first whatever is the ordering in the file itself.

Reply:
Changed the code to parse constants first.

Reply:
The order is insertion order guaranteed using and `OrderedDict`. and I wrote a little test for myself, it will store the sections in order, by file. so if file1 has a `[binaries]` section, and file2 has a `[constants]` section and a `[binaries]` section, `[binaries]` will be iterated before `[constants]` I think it's going to be an endless source of headaches to not guarantee that `[constants]` will be read first.

Reply:
Ok, thanks for confirming what I though. So my updated patch is the only reliable way I think. Right?